_posts/drafts/go-object-model.md

```diff
@@ -0,0 +1,286 @@
+functions in variables, pass them as arguments to other functions,
+and even pass a function directly as an argument too.
+
+With this example in mind we have the opportunity to start exploring
+Go's objects. Can a method satisfy the **Adder** type ? Depending
+on your background this may sound counter intuitive,
+lets take a look at an adder object:
+
+```go
+type ObjectAdder struct{}
+
+func (o *ObjectAdder) Add(a int, b int) int {
+	return a + b
+}
+```
+
+Sound about right, adding it on our example:
+
+```go
+package main
+
+import "fmt"
+
+type Adder func(int, int) int
+
+// Same type as Adder
+func add(a int, b int) int {
+	return a + b
+}
+
+func abstractedAdd(a Adder, b int, c int) int {
+	return a(b, c)
+}
+
+type ObjectAdder struct{}
+
+func (o *ObjectAdder) Add(a int, b int) int {
+	return a + b
+}
+
+func main() {
+	var a Adder
+	fmt.Printf("Adder: %v\n", a)
+	a = add
+	fmt.Printf("Adder initialized: %v\n", a)
+	fmt.Printf("func: %d + %d = %d\n", 1, 1, abstractedAdd(a, 1, 1))
+	fmt.Printf("func: %d + %d = %d\n", 1, 1, abstractedAdd(add, 1, 1))
+
+	var o *ObjectAdder
+	fmt.Printf("object: %d + %d = %d\n", 1, 1, abstractedAdd(o.Add, 1, 1))
+}
+```
+
+Result:
+
+```
+Adder: <nil>
+Adder initialized: 0x401000
+func: 1 + 1 = 2
+func: 1 + 1 = 2
+object: 1 + 1 = 2
+```
+
+Yep, it worked. Differently from interfaces, the function signature will not
+match any kind of method name, you just pass the method as a parameters, since
+the method is actually just a function, it could be something like:
+
+```go
+	var o *ObjectAdder
+	fmt.Printf("object: %d + %d = %d\n", 1, 1, abstractedAdd(o.Whatever, 1, 1))
+```
+
+Should work just fine. Not convinced that the method is just a function ?
+Let's add this:
+
+```go
+	fmt.Printf("func add: %T\n", add)
+	fmt.Printf("object.Add: %T\n", o.Add)
+```
+
+Result:
+
+```
+func add: func(int, int) int
+object.Add: func(int, int) int
+```
+
+Do you see any difference on the type of the free function and
+the object method ? No...because there is none. That is why passing
+it as the parameter worked. This also explains another thing going on
+on our code that sometimes makes newcomers to Go confused.
+
+There is no fully initialized ObjectAdder on our example. I used a pointer
+by purpose, as you can see the pointer is not initialized at all (it is nil),
+yet it worked.  In any other object oriented language that I know (not much,
+C++, Python and Java basically) this would never work, but in
+Go it worked, why ?
+
+Well because in Go, there is no methods at all, there is no method type,
+methods are actually syntactic sugar for calling functions passing the
+struct as the first parameter (as people are used to do in C).
+
+Not convinced ? Lets elaborate our example:
+
+```go
+	fmt.Printf("ObjectAdder.Add: %T\n", (*ObjectAdder).Add)
+	fmt.Printf("ObjectAdder.Add: %d + %d = %d\n", 1, 1, (*ObjectAdder).Add(nil, 1, 1))
+```
+
+Result:
+
+```
+ObjectAdder.Add: func(*main.ObjectAdder, int, int) int
```

Comment:
This confuses me a lot, because o.Add hide the object being passed as first argument but (*ObjectAdd).Add make it explicit. Maybe there's room to explain this here.